src/basic/process-util.c

```diff
@@ -275,21 +276,59 @@ int get_process_cmdline(pid_t pid, size_t max_length, bool comm_fallback, char *
 }
 
 void rename_process(const char name[8]) {
+        static char *mm = NULL;
+        static size_t mm_size = 0;
+        size_t l;
+        char *old_mm = NULL;
+        size_t old_mm_size = 0;
+
+        /* This is a like a poor man's setproctitle(). It changes the comm field, argv[0], and also the glibc's
+         * internally used name of the process. For the first one a limit of 16 chars applies; to the second one in
+         * many cases one of 10 (i.e. length of "/sbin/init") — however if we have CAP_SYS_RESOURCES it is unbounded;
+         * to the third one 7 (i.e. the length of "systemd". If you pass a longer string it will likely be
+         * truncated. */
+
         assert(name);
 
-        /* This is a like a poor man's setproctitle(). It changes the
-         * comm field, argv[0], and also the glibc's internally used
-         * name of the process. For the first one a limit of 16 chars
-         * applies, to the second one usually one of 10 (i.e. length
-         * of "/sbin/init"), to the third one one of 7 (i.e. length of
-         * "systemd"). If you pass a longer string it will be
-         * truncated */
+        l = strlen(name);
 
+        /* First step, change the comm field. */
         (void) prctl(PR_SET_NAME, name);
 
+        /* Second step, change glibc's ID of the process name. */
         if (program_invocation_name)
                 strncpy(program_invocation_name, name, strlen(program_invocation_name));
 
+        /* Third step, allocate some memory we can place the new name in. To be usable for PR_SET_MM_ARG_START this
+         * needs to be memory suitable for a stack, hence we allocate it with MAP_GROWSDOWN. */
+        if (mm_size < l+1) {
+                size_t nn_size;
+                char *nn;
+
+                nn_size = PAGE_ALIGN(l+1);
+                nn = mmap(NULL, nn_size, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_GROWSDOWN, -1, 0);
+                if (nn == MAP_FAILED)
+                        goto fallback;
+
+                old_mm = mm;
+                old_mm_size = mm_size;
+
+                mm = nn;
+                mm_size = nn_size;
+        }
+
+        strncpy(mm, name, mm_size);
+
+        if (prctl(PR_SET_MM, PR_SET_MM_ARG_START, (unsigned long) mm, 0, 0) < 0)
+                goto fallback;
```

Comment:
> Given that this error path is unlikely to be hit

BTW, this error path is likely to be hit inside the user namespace (i.e. `systemd-nspawn -U`):
```
...
32    prctl(PR_SET_NAME, "(sh)\0\177\0\0@4\245\331\276U\0\0") = 0
32    mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_GROWSDOWN, -1, 0) = 0x7f2fcf80b000
32    prctl(PR_SET_MM, PR_SET_MM_ARG_START, 0x7f2fcf80b000, 0, 0) = -1 EPERM (Operation not permitted)
```